### PR TITLE
Fix checkout with quantity zero

### PR DESCRIPTION
**Pull request by dkubb**

### CODE REVIEW

File: .buildkite/pipeline.yml [hunk 1]
```diff
@@ -109,12 +109,13 @@ steps:
       queue: mrh-build
     timeout_in_minutes: 10
 
-  - name: ':skull: Spree::OrdersController#show'
+  - name: ':skull: Spree::OrdersController#{show,update}'
```
Comment (mbj) [inline .buildkite/pipeline.yml]: Should we whitelist all methods on the `Spree::OrdersController`?

File: core/app/models/spree/order.rb [hunk 2]
```diff
@@ -17,7 +17,7 @@ class Order < Spree::Base
     # Error raised on concurrent order operations
     OrderBusyError = Class.new(RuntimeError)
 
-    state_machine :state, initial: :cart, use_transactions: false, action: :save! do
+    state_machine :state, initial: :cart, use_transactions: false, action: :save do
```
Comment (mbj) [inline core/app/models/spree/order.rb]: I'd love to get a rid of the state machine triggered saves soon. Not this PR.

File: frontend/app/controllers/spree/orders_controller.rb [hunk 3]
```diff
@@ -24,19 +24,21 @@ def show
...
-              @order.next if @order.cart?
-              redirect_to checkout_state_path(@order.checkout_steps.first)
-            else
-              redirect_to cart_path
-            end
+      checkout = params.has_key?(:checkout)
+
+      if !@order.contents.update_cart(order_params) ||
+        (checkout && @order.cart? && !@order.next)
+        return render :edit, status: :unprocessable_entity
```
Comment (mbj) [inline frontend/app/controllers/spree/orders_controller.rb]: Lets add parenthesis around the `render` call?

File: .buildkite/pipeline.yml (hunk 1 above)
Comment (mbj) [inline .buildkite/pipeline.yml]: > If so, then I agree, this is a good idea.

If you do not touch such methods, you should not have to add `--ignore-subject`. IMO we should never add subject ignores from now on.

Comment (dkubb) [inline .buildkite/pipeline.yml]: Oh, so you're saying that in the future if someone touches one of those other methods they'll be forced to fix the mutation coverage. I agree with this.

File: core/app/models/spree/order.rb (hunk 2 above)
Comment (dkubb) [inline core/app/models/spree/order.rb]: @mbj do you see any problems with changing this from `#save!` to `#save`? I can't think of any places where we do `Spree::Order#save!`. We should only be using `#save` and testing the return value.

Comment (dkubb) [inline core/app/models/spree/order.rb]: > I think that the return of the action is swallowed. 

I'm not sure what you mean by this? The return value of the `#save` is returned from `@order.next`.

Comment (mbj) [inline core/app/models/spree/order.rb]: k

File: frontend/app/controllers/spree/orders_controller.rb [hunk 4]
```diff
@@ -24,19 +26,23 @@ def show
...
+        format.html do
+          url =
+            if checkout
+              checkout_state_path(INITIAL_STATE)
             else
-              redirect_to cart_path
+              cart_path
             end
-          end
+          redirect_to(url)
```
Comment (mbj) [inline frontend/app/controllers/spree/orders_controller.rb]: I wounder if the assignment to `url` should be flattened into:

```
redirect_to(checkout ? checkout_state_path(INITIAL_STATE) : cart_path)
```

File: frontend/spec/controllers/spree/orders_controller_spec.rb [hunk 5]
```diff
@@ -134,41 +134,113 @@ def do_request
...
-        it "should render the edit view (on failure)" do
-          spree_put :update, {
+        let(:attributes) do
+          {
             'order' => {
               'line_items_attributes' => {
                 '0' => {
                   'id'       => line_item.id.to_s,
-                  'quantity' => Spree::LineItem::QUANTITY_RANGE.max.succ.to_s
+                  'quantity' => quantity.to_s
```
Comment (mbj) [inline frontend/spec/controllers/spree/orders_controller_spec.rb]: Should we use symbol keys here and `#deep_stringify_keys` on the `spree_post` call side?

**Comment by mbj** [comment]
@dkubb I added 2 smaller questions.

File: frontend/spec/controllers/spree/orders_controller_spec.rb (hunk 5 above)
Comment (dkubb) [inline frontend/spec/controllers/spree/orders_controller_spec.rb]: Sure. That should result in easier-to-read test setup.

File: frontend/app/controllers/spree/orders_controller.rb (hunk 4 above)
Comment (dkubb) [inline frontend/app/controllers/spree/orders_controller.rb]: If this fits without wrapping I'll make this change.

Comment (mbj) [inline frontend/app/controllers/spree/orders_controller.rb]: I'd also be fine with:

```
redirect_to(
  checkout ? checkout_state_path(INITIAL_STATE) : cart_path
)
```

Comment (mbj) [inline frontend/app/controllers/spree/orders_controller.rb]: Also we could define a method:

```
def checkout?
  params.key?(:checkout)
end

def redirect_checkout_path
  checkout ? checkout_state_path(INITIAL_STATE) : cart_path
end
```

To remove both the lvar assignemnt to `checkout` and the need to calculate the path in the already big `#update`.

File: frontend/app/controllers/spree/orders_controller.rb [hunk 6]
```diff
@@ -24,19 +26,23 @@ def show
     end
 
     def update
-      if @order.contents.update_cart(order_params)
-        respond_with(@order) do |format|
-          format.html do
-            if params.has_key?(:checkout)
-              @order.next if @order.cart?
-              redirect_to checkout_state_path(@order.checkout_steps.first)
+      checkout = params.has_key?(:checkout)
```
Comment (mbj) [inline frontend/app/controllers/spree/orders_controller.rb]: Do we need `#has_key?` here or can we use the more canonical `#key` ?

Comment (dkubb) [inline frontend/app/controllers/spree/orders_controller.rb]: Good catch. We should be able to use `#key?` here.

File: frontend/app/controllers/spree/orders_controller.rb [hunk 7]
```diff
@@ -24,19 +26,17 @@ def show
...
+      checkout = params.key?(:checkout)
+
+      if !@order.contents.update_cart(order_params) ||
+        (checkout && @order.cart? && !@order.next)
+        return render(:edit, status: :unprocessable_entity)
+      end
+
+      respond_to do |format|
+        format.html do
+          redirect_to(checkout ? checkout_state_path(INITIAL_STATE) : cart_path)
```
Comment (mbj) [inline frontend/app/controllers/spree/orders_controller.rb]: Bump for: https://github.com/MountainRoseHerbs/spree/pull/537#discussion_r85846715

File: frontend/spec/controllers/spree/orders_controller_spec.rb [hunk 8]
```diff
@@ -134,41 +134,113 @@ def do_request
...
+        spree_put(:update, attributes.deep_stringify_keys)
+      end
+
       context "with authorization" do
         before do
-          allow(controller).to receive_messages current_order: order
+          allow(controller).to receive_messages(current_order: order)
+
+          allow(order).to receive_messages(state: state, next: true)
+          allow(order).to receive(:cart?) { state.eql?('cart') }
```
Comment (mbj) [inline frontend/spec/controllers/spree/orders_controller_spec.rb]: Can the value for `cart?` be evaluated outside the lambda via the `receive_messages` above?

**Pull request merged**